Add --page filter to spray command for targeted pointer spraying

Eshaan-byte opened the pull request:
Adds a `--page` filter to the `spray` command that allows filtering pointers to only spray those pointing to a specific vmmap page.

## Changes
- Added `--page` argument that accepts both page names (e.g., `[heap]`, `[stack]`) and addresses (hex or decimal)
- The filter works independently or in combination with the existing `--only-funcptrs` flag
- Provides contextual feedback messages based on which filters are active

## Usage Examples
```bash
# Spray only pointers pointing to the heap
spray 0x555555558000 --page=[heap]

# Spray only pointers pointing to a specific address/page
spray 0x555555558000 --page=0x7ffff7dd5000

# Combine with --only-funcptrs to spray only function pointers pointing to heap
spray 0x555555558000 --only-funcptrs --page=[heap]

Eshaan-byte committed:
Add --page filter to spray command for targeted pointer spraying

Implements the --page flag that allows filtering pointers to only spray
those pointing to a specific vmmap page. The flag accepts both page names
(e.g., [heap], [stack]) and addresses (hex or decimal). This feature works
in combination with the existing --only-funcptrs flag.

Fixes #3458

disconnect3d commented on pwndbg/commands/spray.py:
```diff
@@ -29,22 +29,53 @@
+            if page.startswith("0x"):
+                page_addr = int(page, 16)
+            else:
+                page_addr = int(page)
```
`page_addr = int(page, 16) if page.startswith("0x") else int(page)`

disconnect3d commented on pwndbg/commands/spray.py:
```diff
@@ -29,22 +29,53 @@
+                page_addr = int(page, 16)
+            else:
+                page_addr = int(page)
+            target_page = pwndbg.aglib.vmmap.find(page_addr)
```
this shouldnt be in try/except.

disconnect3d commented on pwndbg/commands/spray.py:
```diff
@@ -29,22 +29,53 @@
+            # Find the page by name
+            for p in pwndbg.aglib.vmmap.get():
+                if p.objfile == page:
+                    target_page = p
```
This may allow for multiple pages. e.g. "libc" will account for multiple libc pages. We probably want to have some additional filtering here for page[idx] or some pages permissions filter ehh

disconnect3d reviewed (changes requested):
See comments

disconnect3d commented:
Btw for page filter, u probably want partial match, e.g. `-p=heap` should match `[heap]` page ojbfile.

Eshaan-byte committed:
Update autogenerated documentation for spray command

Eshaan-byte commented:
Hi @disconnect3d! I've addressed all the review feedback:

The implementation now supports partial matching while handling the case where multiple pages match (e.g., "libc" matching multiple libc pages). When multiple matches are found, it displays all options with their address ranges and prompts the user to use a more specific page name or an address.

disconnect3d commented on pwndbg/commands/spray.py:
```diff
@@ -29,22 +29,64 @@
+                print(M.error(f"Multiple pages match '{page}':"))
+                for idx, p in enumerate(matching_pages):
+                    print(f"  [{idx}] {p.objfile} ({hex(p.start)}-{hex(p.end)})")
+                print(M.notice("Please specify the exact page name or use an address"))
```
While I think it makes sense for the user to specify this, what if the user wants all of the pages with such a name instead of filtering an exact specific one?

disconnect3d commented:
@Eshaan-byte Please see my comment/question. We probably want to handle that case as well. We also need to fix conflicts from recent changes on the dev branch

Eshaan-byte commented:
@disconnect3d Good point! I've updated the implementation to handle the case where users want all matching pages. New behavior:
When multiple pages match (eg, --page=libc matching multiple libc pages), it now sprays pointers to any of those matching pages instead of erroring out
Displays a notice: "Found N pages matching 'X', will spray pointers to any of them"

Output messages adapt based on whether filtering a single page or multiple pages

Also resolved the merge conflicts with the latest dev branch. Changes are in the latest commit d01e26d58.

disconnect3d commented on pwndbg/commands/spray.py:
```diff
@@ -66,18 +108,68 @@ def spray(addr, length, value, only_funcptrs) -> None:
+
+                # Check if pointer points to executable page (if --only-funcptrs is set)
+                if only_funcptrs:
+                    should_spray = should_spray and (ptr_page is not None and ptr_page.execute)
```
Why the `should_spray and` part if `should_psray` is always set to True before that?

disconnect3d replied:
Please read the code yourself before sending it from an AI :(

disconnect3d commented:
@Eshaan-byte ping?

disconnect3d commented:
@Eshaan-byte are u going to update this or should we close it

disconnect3d commented:
Closing due to inactivity
